PHP Chainguard Image Guide (#270)
chainguard-dev/edu: content/chainguard/chainguard-images/reference/php/getting-started.md

---
title: "Getting Started with the PHP Chainguard Images"
type: "article"
description: "Documentation on how to get started with the PHP Chainguard Images"
date: 2023-01-09T11:07:52+02:00
lastmod: 2023-01-09T11:07:52+02:00
draft: false
images: []
referenceimage: php
menu:
  docs:
    parent: "php-guides"
weight: 10
toc: true
---

The PHP images maintained by Chainguard are a mix of development and production distroless images that are suitable for building and running command-line PHP workloads.

> Because PHP applications typically require user-land package installs via Composer, using a pure distroless image for building your application would not work. In cases like this, you'll need to implement a [multi-stage Docker build]() that uses one of the `-dev` images to set up the application.

Reviewer: Missing URL here at `[multi-stage Docker build]()`?

Then, we'll create a separate stage for the final image, which will be based on the `glibc` variant. That is a distroless image also based on Wolfi, which means it doesn't come with Composer or even a shell.

Create a Dockerfile with:

```shell
touch Dockerfile
```

The following Dockerfile will:

1. Start a new build stage based on the `php:latest-glibc-dev` image and call it `builder`;
2. Copy files from the current directory to the `/app` location in the container;
3. Enter the `/app` directory and run `composer install` to install any dependencies;
4. Start a new build stage based on the `php:latest-glibc`;
5. Copy the application from the `builder` stage;
6. Set up the application as entry point for this image.

Copy this content to your own `Dockerfile`:

```Dockerfile
FROM cgr.dev/chainguard/php:latest-glibc-dev AS builder
COPY guides /app
RUN cd /app && \
    composer install --no-progress --no-dev --prefer-dist

FROM cgr.dev/chainguard/php:latest-glibc
COPY --from=builder /app /app

ENTRYPOINT [ "php", "/app/namegen" ]
```
Save the file when you're finished.

You can now build the image with:

```shell

> docker build . -t php-namegen get

Author: oooh sorry this was a typo!!! I'm fixing it now.

```

Once the build is finished, run the image with:

```shell
docker run --rm php-namegen get
```

And you should get output similar to what you got before, with a random name combination.

```
fortuitous-octopus
```

> 

Reviewer: Maybe can add a closing sentence here.

---
title: "Getting Started with the PHP Chainguard Images"
type: "article"
description: "Documentation on how to get started with the PHP Chainguard Images"
date: 2023-01-09T11:07:52+02:00
lastmod: 2023-01-09T11:07:52+02:00
draft: false
images: []
referenceimage: php
menu:
  docs:
    parent: "php-guides"
weight: 10
toc: true
---

The PHP images maintained by Chainguard are a mix of development and production distroless images that are suitable for building and running command-line PHP workloads.

> Because PHP applications typically require user-land package installs via Composer, using a pure distroless image for building your application would not work. In cases like this, you'll need to implement a [multi-stage Docker build](https://docs.docker.com/build/building/multi-stage/) that uses one of the `-dev` images to set up the application.

Reviewer: "user-land" is a new word for me 😄

Author: I am going to change this 😄 it can be confusing, I am not sure everyone will understand it as I do, lol. but I meant dependencies that are not system-wide

{{< /details >}}

## Step 1: Setting up a Demo Application

We'll start by creating a basic command-line PHP application to serve as a demo. This app will generate random names based on a list of nouns and adjectives. To exemplify usage with Composer, the app will have a single dependency on [minicli](https://github.com/minicli/minicli), a minimalist CLI framework for PHP.

First, create a directory for your app. In this guide we'll use `wolfi-php`:

```shell
mkdir ~/wolfi-php && cd ~/wolfi-php
```

If you have a local PHP development environment with Composer, you can run the following command to download the single dependency of this app.

```
composer require minicli/minicli
```

If you don't have a local PHP development environment, you can use the `php:latest-glibc-dev` image variant with a volume in order to install application dependencies with Composer:

```shell
docker run --rm -v ${PWD}:/work --entrypoint composer \
    cgr.dev/chainguard/php:latest-glibc-dev \
    require minicli/minicli --working-dir=/work
```

> If you used the Docker method, make sure permissions are right on the generated files:

Reviewer: ```suggestion
If you used the Docker method, make sure permissions are set correctly on the generated files:
```

```shell
sudo chown -R ${USER}.${USER} .
```

> Create a new file to serve as the application entry point. We'll call it "namegen":

Reviewer: What do you think about backtics here?

```suggestion
Create a new file to serve as the application entry point. We'll call it `namegen`:
```

```shell
touch namegen
```

Next, open the file in your code editor of choice, for example with `nano`:

```shell
nano namegen
```

> The following PHP script defines a minimalist CLI app with a single command called "get". It returns a random name based on a list of nouns and a list of adjectives.

Reviewer: ```suggestion
The following PHP script defines a minimalist CLI app with a single command called `get`. It returns a random name based on a list of nouns and a list of adjectives.
```

Create a Dockerfile with:

```shell
touch Dockerfile
```

Then open this file in your code editor of choice, for example `nano`:

```shell
nano Dockerfile
```
The following Dockerfile will:

1. Start a new build stage based on the `php:latest-glibc-dev` image and call it `builder`;
2. Copy files from the current directory to the `/app` location in the container;
3. Enter the `/app` directory and run `composer install` to install any dependencies;
4. Start a new build stage based on the `php:latest-glibc`;
5. Copy the application from the `builder` stage;
6. Set up the application as entry point for this image.

Copy this content to your own `Dockerfile`:

```Dockerfile
FROM cgr.dev/chainguard/php:latest-glibc-dev AS builder
COPY guides /app
RUN cd /app && \
    composer install --no-progress --no-dev --prefer-dist

FROM cgr.dev/chainguard/php:latest-glibc
COPY --from=builder /app /app

ENTRYPOINT [ "php", "/app/namegen" ]
```
Save the file when you're finished.

You can now build the image with:

```shell

> docker build . -t php-namegen

Reviewer: I am still getting an issue with this command

On Debian:

```
Sending build context to Docker daemon  296.4kB
Step 1/6 : FROM cgr.dev/chainguard/php:latest-glibc-dev AS builder
 ---> f06865c7bb16
Step 2/6 : COPY guides /app
COPY failed: file not found in build context or excluded by .dockerignore: stat guides: file does not exist
```

on macOS:

```
[+] Building 0.1s (8/10)                                                        
 => [internal] load build definition from Dockerfile                       0.0s
 => => transferring dockerfile: 303B                                       0.0s
 => [internal] load .dockerignore                                          0.0s
 => => transferring context: 2B                                            0.0s
 => [internal] load metadata for cgr.dev/chainguard/php:latest-glibc       0.0s
 => [internal] load metadata for cgr.dev/chainguard/php:latest-glibc-dev   0.0s
 => [internal] load build context                                          0.0s
 => => transferring context: 2B                                            0.0s
 => CACHED [builder 1/3] FROM cgr.dev/chainguard/php:latest-glibc-dev      0.0s
 => CACHED [stage-1 1/2] FROM cgr.dev/chainguard/php:latest-glibc          0.0s
 => ERROR [builder 2/3] COPY guides /app                                   0.0s
------
 > [builder 2/3] COPY guides /app:
------
failed to compute cache key: "/guides" not found: not found
```

Author: oh I see the problem, it is really strange I have no idea how this "guides" showed up in the Dockerfile, it was probably a faulty find and replace. Sorry! I just updated it now.

            ]
        },
        "Metadata": {
            "LastTagTime": "2023-01-10T19:02:13.062958609+01:00"
        }
    }
]

```
In such cases, the last `FROM` section from the Dockerfile is the one that composes the final image. That's why in our case it only adds one layer on top of the base `php:latest-glibc` image, containing the `COPY` command we use to copy the application from the build stage to the final image.

It's worth highlighting that nothing is carried from one stage to the other unless you copy it. That facilitates creating a slim final image with only what's necessary to execute the application.

## Advanced Usage

If your use case requires a more specific set of packages that the general-purpose PHP Chainguard Images don't provide, you'll first need to check if the package you need is available for the base distro you choose.

For Wolfi-based images, check the [Wolfi](https://github.com/wolfi-dev/os) distro repository. If the package you want is there, you can use the [wolfi-base](/chainguard/chainguard-images/reference/wolfi-base/overview) image in a Dockerfile and install what you need with `apk`, then use the resulting image as base for your app.

> For Alpine-based images, if the packages you need are available in the [Alpine](https://pkgs.alpinelinux.org/packages) repositories, you can use the [alpine-base](/chainguard/chainguard-images/reference/alpine-base/overview) image in a Dockerfile and install what you need with `apk`, then use the resulting image as base for your app.

Reviewer: ```suggestion
For Alpine-based images, if the packages you need are available in the [Alpine](https://pkgs.alpinelinux.org/packages) repositories, you can use the [alpine-base](/chainguard/chainguard-images/reference/alpine-base/overview) image in a Dockerfile and install what you need with `apk`, then use the resulting image as the base for your app.
```

> If the packages you need are not available for your your base distro of choice, you can build your own apks using [melange](/open-source/melange/overview). Please refer to [this guide](/open-source/melange/tutorials/getting-started-with-melange/) for more information on how to do that.

Reviewer: I think just ending the sentence earlier will make it a little more formal

```suggestion
If the packages you need are not available for your your base distro of choice, you can build your own apks using [melange](/open-source/melange/overview). Please refer to [this guide](/open-source/melange/tutorials/getting-started-with-melange/) for more information.
```
